crates/doc/src/shape/widen.rs

```diff
@@ -2,232 +2,393 @@
-        // Otherwise, we need to merge their shapes into `additionalProperties`.
-        if !new_fields.is_empty() {
-            // additionalProperties: false
-            if additional_properties.type_.eq(&types::INVALID) {
-                // These new shapes can not conflict with existing properties by definition
-                // because they were produced by the right-hand-side of the `merge_join_by`.
-                // That is, these fields explicitly do not yet exist on this shape.
-                self.properties.extend(new_fields.into_iter());
-                self.properties.sort_by(|a, b| a.name.cmp(&b.name))
-            } else {
-                for field in new_fields {
-                    additional_properties =
-                        Shape::union(additional_properties.clone(), field.shape);
+                match &mut self.additional_properties {
+                    // If `additionalProperties` is unset, its default is the "true" schema
+                    // which accepts any JSON document. No need to widen further.
+                    None => return false,
+                    // If `additionalProperties` is an explicit schema _other_ than "false", widen it.
+                    Some(additional) if additional.type_ != types::INVALID => {
+                        return additional.widen(rhs.value());
+                    }
+                    // `additionalProperties` is "false". Fall through to add a new property.
+                    _ => (),
                 }
+
+                let mut shape = Shape::nothing();
+                _ = shape.widen(rhs.value());
+
+                new_fields.push(ObjProperty {
+                    name: rhs.property().into(),
+                    // A field can only be required if every single document we've seen
+                    // has that field present. This means that ONLY fields that exist
+                    // on the very first object we encounter for a particular location should
+                    // get marked as required, as any subsequent "new" fields
+                    // by definition did not exist on previous objects (and so cannot be required)
+                    // otherwise they would already be in the Shape
+                    // (and we would be in the `EoB::Both` branch).
+                    is_required: is_first,
+                    shape,
+                });
+
+                true
+            };
+
+        let mut changed = is_first;
+        let mut lhs_it = self.properties.iter_mut();
+        let mut rhs_it = fields.iter();
+        let mut maybe_lhs = lhs_it.next();
+        let mut maybe_rhs = rhs_it.next();
+
+        // Perform an ordered merge over `lhs_it` and `rhs_it`.
+        // This loop structure is much faster than Itertools::merge_join_by.
+        loop {
+            match (&mut maybe_lhs, &maybe_rhs) {
+                (Some(lhs), Some(rhs)) => match lhs.name.as_ref().cmp(rhs.property()) {
```

Comment:
This is relying deeply on the assumption that lhs and rhs are both sorted, and anything that results in them not being sorted would cause subtly wrong behavior. I wonder if it would be worth adding a check here to ensure that `lhs < lhs_it.next()` and fail fast if not. 

I'm guessing you don't want to because of performance though, so... 🤞 we don't accidentally create incorrectly-sorted properties/fields.